Introduce Grey Image

Vici37 opened the pull request:
Introducing the grey image type as the first step to feature extraction from images. I had meant to introduce bit masking and an initial threshold operation, but felt this change had already grown large enough to get eyes upon.

Some of the changes I made I think also help address #12 .

Vici37 commented on spec/pluto/grey_image_spec.cr:
```diff
@@ -0,0 +1,63 @@
+  Digest::SHA1.hexdigest(content)
+end
+
+describe Pluto::GreyImage do
```
I didn't want to add the dependency without asking here first: what are thoughts around [spectator](https://gitlab.com/arctic-fox/spectator)? I've enjoyed using it for personal projects and find it provides a lot of testing nice-to-haves (such as the `expect` macro which shows what's being run as well as what the differences were, rather than the current `should` that only shows the differences).

sanks64 replied:
Well, there would have to be a reason to introduce Spectator. I'm not sure if showing what's being run is a good one (unless I'm misunderstanding).

Vici37 replied:
I don't think spectator introduces anything as a must have, but it does also have a nice [suite of features](https://gitlab.com/arctic-fox/spectator/-/wikis/home) that I find makes testing more streamlined. 

The specific bit I was thinking about is mentioned [here](https://gitlab.com/arctic-fox/spectator/-/tree/master#assertions), but I think regular `should`s now show the expression that failed, not just the result.

Regardless, sounds like spectator isn't a good fit at this time :grin:

sanks64 replied:
The features look good, but yeah, I think it's better to prioritize simplicity for now. :)

Vici37 commented on src/pluto/format/jpeg.cr:
```diff
@@ -41,13 +45,27 @@ module Pluto::Format::JPEG
-        string.write_byte(@blue.unsafe_fetch(index))
+        string.write_byte(red.unsafe_fetch(index))
+        string.write_byte(green.unsafe_fetch(index))
+        string.write_byte(blue.unsafe_fetch(index))
```
Now grey images can write as if they're regular images :)

Vici37 commented on src/pluto/format/open.cr:
```diff
@@ -0,0 +1,9 @@
+module Pluto::Format::Open
```
I put this into the `Format` namespace even though it behaves more like an `Operator`, since it directly uses the other formats. Let me know if this is an issue.

Provides a nice convenience method for opening up images.

sanks64 replied:
Let's not add these convenience methods as they are simple enough to implement yourself and otherwise just increase complexity of the library.

Vici37 replied:
I'm assuming the most common way of creating `Image` objects with this library will be through file reading and writing files. I'd rather that common complexity be taken on by the library than forcing all users of the library to re-implement the same logic. Would it be alright if we kept this and the `Save` modules?

I can remove the specific `open_ppm` and `open_jpeg` methods and handle the `File.open` stuff in this module instead?

sanks64 replied:
The thing is, I'm afraid it will only bring clutter because:

1. There are multiple ways to do the same thing. When a user sees the public API, they are presented with a bad choice: to use the convenience methods or not. What if they decide not to? It sounds like a mess.
2. You still have to use `IO` in some cases. For example, in one of my projects, images come from [grok.cr](https://github.com/phenopolis/grok.cr), so should I use `IO` directly in this case and the convenience methods in others? Again, this sounds like a mess.
3. We have to remember to update these methods each time a new format is added to the library.

sanks64 replied:
Alternatively, you can create an extension and we will add a link to it in `README.md`.

Vici37 replied:
Removed the `Open` and `Save` operators and methods.

I'm hoping your arguments are against a fictional list of poorly implemented "convenience" methods that clutter up the API, rather than this `open` and `save` methods directly :/ 

1. I'm not sure how presenting a user with the choice of convenience methods leads to a mess - assuming the user is intelligent (especially if they're a developer), they'll use the method that most closely matches their use case. It wouldn't be a "bad" choice, it would be "their" choice.
2. For your grok.cr use case, I'd expect you to use whichever method matches your use case the best. If you were reading or writing files primarily, these `open` and `save` methods would probably be a good convenience for you. If you were reading / writing to a socket directly, the `IO` based methods would be what you'd use. Again, I'm not seeing the mess for this specific set of methods.
3. Sure, but there's always a risk of introducing bugs when adding a new format. The type of bug for missing to add this to the convenience methods is a pretty low risk / low consequence bug and easily fixed through a PR.

Regardless, I removed the methods, so there is no concern for scope creep this way.

Vici37 commented on src/pluto/format/save.cr:
```diff
@@ -0,0 +1,10 @@
+module Pluto::Format::Save
```
Ditto as the `Open` module. Provides nice convenience method for saving an image to file.

Vici37 commented on src/pluto/grey_image.cr:
```diff
@@ -0,0 +1,76 @@
+
+  def blue : Array(UInt8)
+    @grey
+  end
```
These methods help make grey images completely compatible with regular color images (for the sake of saving).

Vici37 commented on src/pluto/operation/bilinear_resize.cr:
```diff
@@ -1,9 +1,9 @@
 module Pluto::Operation::BilinearResize
-  def bilinear_resize(width : Int32, height : Int32) : Image
+  def bilinear_resize(width : Int32, height : Int32) : self
```
This and below changes allow `image.bilinear_resize(...)` and `grey_image.bilinear_resize(...)` to return their respective types (with `image` being an instance of `Image` and `grey_image` being an instance of `GreyImage`).

sanks64 reviewed (commented):
It seems to me that there's a fundamental issue with how this is being approached.

In the current implementation, `GreyImage` is being added as an afterthought. Since Pluto still isn't 1.0.0, I think it's okay to break things.

One cleaner approach that I would like is introducing a parent class (`Image`?) and then inheriting from that (`RGBImage` and `GreyImage`/`GrayscaleImage`?).

Is this something you can explore?

Vici37 commented:
Can do! I had a common parent class in an earlier iteration but misattributed an error I saw as a result of that and reverted, and then didn't re-introduce it after I fixed the real cause.

I also wasn't sure what your feeling would be on changing that hierarchy, so glad you suggested it :smile:

Vici37 commented on src/pluto/image.cr:
```diff
@@ -2,67 +2,43 @@ require "./bindings/*"
-    in ChannelType::Alpha then @alpha
-    end
-  end
+  forward_to_rgb_image open, from_ppm, from_jpg
```
Added this macro for backwards compatibility when changing from `Image` -> `RGBImage`.

sanks64 commented on spec/pluto/grey_image_spec.cr:
```diff
@@ -0,0 +1,63 @@
+  it "gaussian blurs" do
+    hash(sample_image.gaussian_blur(10).to_ppm).should eq "df13de316f347c955309abcada06657d00b55bf5"
+  end
+end
```
Let's move the specs to respective files, so that we have something like:

```crystal
require "../../spec_helper"

describe Pluto::Operation::BoxBlur do
  describe "#box_blur" do
    it "works with GreyImage" do
      # ...
    end

    it "works with RGBImage" do
      # ...
    end
  end

  describe "#box_blur!" do
    it "works with GreyImage" do
      # ...
    end

    it "works with RGBImage" do
      # ...
    end
  end
end
```

Vici37 replied:
Updated specs! There's a _lot_ of duplicated code, and I ran out of time coming up with useful methods to dedupe it all :/

sanks64 replied:
Thanks!

The most important part is that specs are readable and coherent. We can always get rid of duplication later.

sanks64 commented on src/pluto/format/jpeg.cr:
```diff
@@ -1,6 +1,10 @@
-    def self.from_jpeg(image_data : String) : Image
+    def self.open_jpeg(filename : String) : self
+      from_jpeg(File.read(filename))
+    end
```
Let's remove such methods as well because they add extra complexity for what can be easily done manually.

sanks64 commented on src/pluto/format/jpeg.cr:
```diff
@@ -41,13 +45,27 @@ module Pluto::Format::JPEG
 
+  def to_jpeg(filename : String, quality : Int32 = 100) : Nil
+    File.open(filename, "w") { |f| to_jpeg(f, quality) }
+  end
```
And this.

sanks64 commented on src/pluto/format/ppm.cr:
```diff
@@ -1,8 +1,10 @@
-      io = IO::Memory.new(image_data)
+    def self.open_ppm(filename : String) : self
+      from_ppm(File.read(filename))
+    end
```
And this.

sanks64 commented on src/pluto/format/ppm.cr:
```diff
@@ -32,18 +34,31 @@ module Pluto::Format::PPM
+
+  def to_ppm(filename : String) : Nil
+    File.open(filename, "w") { |f| to_ppm(f) }
+  end
```
And this.

sanks64 commented on src/pluto/grey_image.cr:
```diff
@@ -0,0 +1,63 @@
+  property width : Int32
+  property height : Int32
+
+  def self.new(red : Array(UInt8), green : Array(UInt8), blue : Array(UInt8), alpha : Array(UInt8), width : Int32, height : Int32)
```
I think it makes sense to remove this method since we already have `RGBImage#to_grey`?

sanks64 replied:
@Vici37

sanks64 commented on src/pluto/grey_image.cr:
```diff
@@ -0,0 +1,63 @@
+
+  def [](channel_type : ChannelType) : Array(UInt8)
+    # All channels are grey. There can only be grey.
+    @grey
```
Can we raise here as we do in `#[]=`?

Vici37 replied:
I initially didn't do that so that this method would behave the same way as the `red`, `green`, and `blue` methods using the `[ChannelType::Red]`, `[ChannelType::Green]`, and `[ChannelType::Blue]` respectively. This was before the common parent class that helped enforce the existence of those previous methods, so it's probably not necessary anymore. Will add in a check :+1:

sanks64 commented on src/pluto/rgb_image.cr:
```diff
@@ -0,0 +1,69 @@
+require "./format/*"
+require "./operation/*"
+
+class Pluto::RGBImage < Pluto::Image
```
Sorry I suggested an incorrect name, but I think it should be `RGBAImage`?

Vici37 replied:
I'm fine with that too. I don't think people would have been that confused without the `A` since there's already an `alpha` method :)

sanks64 commented on src/pluto/grey_image.cr:
```diff
@@ -0,0 +1,63 @@
+require "./image"
+
+class Pluto::GreyImage < Pluto::Image
```
From what I see, the correct name would be `GrayscaleImage`.

https://en.wikipedia.org/wiki/Grayscale

Vici37 replied:
I'm good either way :+1: I don't gain any additional context with the `scale` suffix, but it is technically correct (the best kind of correct).

sanks64 commented on src/pluto/rgb_image.cr:
```diff
@@ -0,0 +1,69 @@
+    end
+  end
+
+  # Convert color image to grey-scale one, using the NTSC formula as default values
```
Please add a dot at the end.

sanks64 commented on src/pluto/rgb_image.cr:
```diff
@@ -0,0 +1,69 @@
+        Math.min(255, red_pix * red_multiplier + @green[i] * green_multiplier + @blue[i] * blue_multiplier).to_u8
+      end,
+      width,
+      height)
```
Let's move this bracket to the next line.
